grudge/op.py

```diff
@@ -403,155 +1002,153 @@ def norm(dcoll, vec, p, dd=None):
+    :returns: an integer denoting the nodal maximum.
+    """
+    # FIXME: Make MPI-aware
+    actx = vec.array_context
+    return reduce(lambda acc, grp_ary: actx.np.maximum(acc, actx.np.max(grp_ary)),
+                  vec, -np.inf)
 
-class _RankBoundaryCommunication:
-    base_tag = 1273
 
-    def __init__(self, dcoll, remote_rank, vol_field, tag=None):
-        self.tag = self.base_tag
-        if tag is not None:
-            self.tag += tag
+def integral(dcoll, vec, dd=None):
+    """Numerically integrates a function represented by a
+    :class:`~meshmode.dof_array.DOFArray` of degrees of freedom.
 
-        self.dcoll = dcoll
-        self.array_context = vol_field.array_context
-        self.remote_btag = BTAG_PARTITION(remote_rank)
+    :arg dcoll: a :class:`grudge.discretization.DiscretizationCollection`.
+    :arg vec: a :class:`~meshmode.dof_array.DOFArray`
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization if not provided.
+    :returns: an integer denoting the evaluated integral.
+    """
 
-        self.bdry_discr = dcoll.discr_from_dd(self.remote_btag)
-        self.local_dof_array = project(dcoll, "vol", self.remote_btag, vol_field)
+    if dd is None:
+        dd = dof_desc.DD_VOLUME
 
-        local_data = self.array_context.to_numpy(flatten(self.local_dof_array))
+    dd = dof_desc.as_dofdesc(dd)
 
-        comm = self.dcoll.mpi_communicator
+    ones = dcoll.discr_from_dd(dd).zeros(vec.array_context) + 1.0
+    return nodal_summation(
+        dcoll, vec * _apply_mass_operator(dcoll, dd, dd, ones)
+    )
 
-        self.send_req = comm.Isend(
-                local_data, remote_rank, tag=self.tag)
+# }}}
 
-        self.remote_data_host = np.empty_like(local_data)
-        self.recv_req = comm.Irecv(self.remote_data_host, remote_rank, self.tag)
 
-    def finish(self):
-        self.recv_req.Wait()
+# {{{  Elementwise reductions
 
-        actx = self.array_context
-        remote_dof_array = unflatten(self.array_context, self.bdry_discr,
-                actx.from_numpy(self.remote_data_host))
+def _map_elementwise_reduction(actx, op_name):
+    @memoize_in(actx, (_map_elementwise_reduction,
+                       "elementwise_%s_prg" % op_name))
+    def prg():
+        return make_loopy_program(
+            [
+                "{[iel]: 0 <= iel < nelements}",
+                "{[idof, jdof]: 0 <= idof, jdof < ndofs}"
+            ],
+            """
+                result[iel, idof] = %s(jdof, operand[iel, jdof])
+            """ % op_name,
+            name="grudge_elementwise_%s_knl" % op_name
+        )
+    return prg()
 
-        bdry_conn = self.dcoll.get_distributed_boundary_swap_connection(
-                dof_desc.as_dofdesc(dof_desc.DTAG_BOUNDARY(self.remote_btag)))
-        swapped_remote_dof_array = bdry_conn(remote_dof_array)
 
-        self.send_req.Wait()
+def elementwise_sum(dcoll, *args):
```

Comment:
I didn't port over `elementwise_{mix,max}` since it's used literally nowhere in the mirgecom stack (or even internally in old world grudge). However, I believe I've written it in such a way that it would be very easy to add them in later if we need them.